lib/phoenix_html/engine.ex

```diff
@@ -16,7 +16,39 @@ defmodule Phoenix.HTML.Engine do
   def init(_opts), do: {:safe, ""}
 
   @doc false
-  def handle_body(body), do: body
+  def handle_body(body) do
+    compact_safe(body)
+  end
+
+  defp compact_safe({:safe, body}) when is_list(body) do
+    case compact_static(List.flatten(body)) do
+      [element] ->
+        {:safe, element}
+      list ->
+        {:safe, list}
+    end
+  end
+  defp compact_safe({:safe, other}) do
+    {:safe, other}
+  end
+  defp compact_safe(other) do
+    other
+  end
+
+  defp compact_static([binary1, binary2 | rest])
+       when is_binary(binary1) and is_binary(binary2) do
+    compact_static([binary1 <> binary2 | rest])
+  end
+  defp compact_static([{_, _, _} = ast | rest]) do
+    compacted = Macro.prewalk(ast, &compact_safe/1)
```

Comment:
No, we only compact inside `{:safe, list}` tuples. But if you had:

```
<%= Foo.bar {:safe, ["foo", "bar"]} %>
```

It would indeed be compacted. The AST traversal was to solve a case of nested "literals", for example:

```
<%= if @foo do %>
  <%= 456 %>
<% end %>
```

would generate something like:

```
case if(fetch!(assigns, :foo), do: {:safe, [["" | "456"] | "\n"]}, else: []) do
  {:safe, safe} -> safe
  binary when is_binary(binary) -> Plug.HTML.escape(binary)
  other -> Phoenix.HTML.Safe.to_iodata(other)
end
```
So we need a way to "get inside" the case. But you're right - it might be enough to do the modifications at the buffer building stage instead of doing them after.